File: crates/ruimte_runtime/src/processes/linux.rs
use std::{
    fs,
    io::Read,
    path::PathBuf,
    time::{SystemTime, UNIX_EPOCH},
};

use anyhow::{Context, anyhow};

use super::sampler::{
    CommandLine, MachineCounters, ProcessSampler, RawProcess, RawSample, parse_nul_command_line,
};

const MAX_COMMAND_LINE_BYTES: u64 = 1024 * 1024;

pub struct LinuxSampler {
    home: PathBuf,
    clock_ticks: u64,
    page_size: u64,
}

impl LinuxSampler {
    pub fn new(home: PathBuf) -> Self {
        let clock_ticks = unsafe { libc::sysconf(libc::_SC_CLK_TCK) }.max(1) as u64;
        let page_size = unsafe { libc::sysconf(libc::_SC_PAGESIZE) }.max(1) as u64;
        Self {
            home,
            clock_ticks,
            page_size,
        }
    }

    fn read_process(&self, pid: u32, boot_seconds: u64) -> anyhow::Result<RawProcess> {
        let base = PathBuf::from("/proc").join(pid.to_string());
        let stat = fs::read_to_string(base.join("stat"))?;
        let (name, fields) = split_process_stat(&stat)?;
        let value = |index: usize| {
            fields
                .get(index)
                .ok_or_else(|| anyhow!("missing proc stat field {index}"))
        };
        let ppid = value(1)?.parse()?;
        let user_ticks: u64 = value(11)?.parse()?;
        let system_ticks: u64 = value(12)?.parse()?;
        let start_ticks: u64 = value(19)?.parse()?;
        let rss_pages: i64 = value(21)?.parse()?;
        let status = fs::read_to_string(base.join("status")).unwrap_or_default();
        let uid = status
            .lines()
            .find_map(|line| line.strip_prefix("Uid:"))
            .and_then(|line| line.split_whitespace().next())
            .and_then(|uid| uid.parse().ok())
            .unwrap_or(u32::MAX);
        let io = fs::read_to_string(base.join("io")).unwrap_or_default();
        let counter = |key: &str| {
            io.lines()
                .find_map(|line| line.strip_prefix(key))
                .and_then(|value| value.trim().parse().ok())
        };
        Ok(RawProcess {
            pid,
            ppid,
            uid,
            start_time: boot_seconds * 1_000_000 + start_ticks * 1_000_000 / self.clock_ticks,
            name,
            path: fs::read_link(base.join("exe"))
                .ok()
                .map(|path| path.to_string_lossy().into_owned()),
            readable: true,
            cpu_ns: Some((user_ticks + system_ticks) * 1_000_000_000 / self.clock_ticks),
            memory: (rss_pages >= 0).then_some(rss_pages as u64 * self.page_size),
            disk_read: counter("read_bytes:"),
            disk_write: counter("write_bytes:"),
        })
    }

    fn boot_seconds() -> anyhow::Result<u64> {
        fs::read_to_string("/proc/stat")?
            .lines()
            .find_map(|line| line.strip_prefix("btime "))
            .and_then(|value| value.parse().ok())
            .ok_or_else(|| anyhow!("missing kernel boot time"))
    }
}

impl ProcessSampler for LinuxSampler {
    fn sample(&self) -> anyhow::Result<RawSample> {
        let at = now_ms();
        let boot = Self::boot_seconds()?;
        let boottime = fs::read_to_string("/proc/uptime")?
            .split_whitespace()
            .next()
            .and_then(|value| value.parse::<f64>().ok())
            .unwrap_or(0.);
        let awake_ms = monotonic_ms().unwrap_or((boottime * 1000.) as u64);
        let processes = fs::read_dir("/proc")?
            .filter_map(Result::ok)
            .filter_map(|entry| entry.file_name().to_string_lossy().parse::<u32>().ok())
            .filter_map(|pid| self.read_process(pid, boot).ok())
            .collect();
        let cpu_line = fs::read_to_string("/proc/stat")?;
        let (total, idle) = cpu_counters(cpu_line.lines().next().unwrap_or_default());
        let memory = fs::read_to_string("/proc/meminfo")?;
        let mem = |key: &str| {
            memory
                .lines()
                .find_map(|line| line.strip_prefix(key))
                .and_then(|line| line.split_whitespace().next())
                .and_then(|value| value.parse::<u64>().ok())
                .map(|value| value * 1024)
        };
        let (disk_free, disk_total) = statvfs(&self.home).unwrap_or((None, None));
        Ok(RawSample {
            at,
            awake_ms,
            asleep_ms: (boottime * 1000.) as u64 - awake_ms.min((boottime * 1000.) as u64),
            processes,
            machine: MachineCounters {
                cores: std::thread::available_parallelism()
                    .map(usize::from)
                    .unwrap_or(1),
                cpu_busy: Some(total.saturating_sub(idle)),
                cpu_total: Some(total),
                memory_used: mem("MemTotal:")
                    .zip(mem("MemAvailable:"))
                    .map(|(total, available)| total.saturating_sub(available)),
                memory_total: mem("MemTotal:").unwrap_or(0),
                disk_free,
                disk_total,
            },
        })
    }

    fn inspect(&self, pid: u32) -> anyhow::Result<Option<(u64, u32)>> {
        match self.read_process(pid, Self::boot_seconds()?) {
            Ok(process) => Ok(Some((process.start_time, process.uid))),
            Err(error)
                if error
                    .downcast_ref::<std::io::Error>()
                    .is_some_and(|error| error.kind() == std::io::ErrorKind::NotFound) =>
            {
                Ok(None)
            }
            Err(error) => Err(error).context("inspect process"),
        }
    }

    fn command_line(&self, pid: u32) -> Option<CommandLine> {
        let base = PathBuf::from("/proc").join(pid.to_string());
        let mut args = Vec::new();
        fs::File::open(base.join("cmdline"))
            .ok()?
            .take(MAX_COMMAND_LINE_BYTES)
            .read_to_end(&mut args)
            .ok()?;
        let mut environment = Vec::new();
        fs::File::open(base.join("environ"))
            .ok()?
            .take(MAX_COMMAND_LINE_BYTES)
            .read_to_end(&mut environment)
            .ok()?;
        let mut line = parse_nul_command_line(&args, None);
        line.env = parse_nul_command_line(&environment, Some(0)).env;
        Some(line)
    }
}

fn split_process_stat(stat: &str) -> anyhow::Result<(String, Vec<&str>)> {
    let open = stat.find('(').ok_or_else(|| anyhow!("invalid proc stat"))?;
    let close = stat
        .rfind(')')
        .ok_or_else(|| anyhow!("invalid proc stat"))?;
    Ok((
        stat.get(open + 1..close).unwrap_or_default().to_owned(),
        stat.get(close + 2..)
            .unwrap_or_default()
            .split_whitespace()
            .collect(),
    ))
}

fn cpu_counters(line: &str) -> (u64, u64) {
    let fields = line
        .split_whitespace()
        .skip(1)
        .filter_map(|value| value.parse::<u64>().ok())
        .collect::<Vec<_>>();
    let total = fields.iter().take(8).sum();
    let idle = fields.get(3).copied().unwrap_or(0) + fields.get(4).copied().unwrap_or(0);
    (total, idle)
}

fn now_ms() -> u64 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap_or_default()
        .as_millis() as u64
}

fn monotonic_ms() -> anyhow::Result<u64> {
    let mut time = libc::timespec {
        tv_sec: 0,
        tv_nsec: 0,
    };
    if unsafe { libc::clock_gettime(libc::CLOCK_MONOTONIC, &mut time) } != 0 {
        return Err(std::io::Error::last_os_error().into());
    }
    Ok(time.tv_sec as u64 * 1000 + time.tv_nsec as u64 / 1_000_000)
}

fn statvfs(path: &std::path::Path) -> anyhow::Result<(Option<u64>, Option<u64>)> {
    use std::ffi::CString;
    use std::os::unix::ffi::OsStrExt;
    let path = CString::new(path.as_os_str().as_bytes())?;
    let mut stat = std::mem::MaybeUninit::<libc::statvfs>::uninit();
    if unsafe { libc::statvfs(path.as_ptr(), stat.as_mut_ptr()) } != 0 {
        return Err(std::io::Error::last_os_error().into());
    }
    let stat = unsafe { stat.assume_init() };
    Ok((
        Some(stat.f_bavail * stat.f_frsize),
        Some(stat.f_blocks * stat.f_frsize),
    ))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_name_without_pid_prefix_and_keeps_embedded_parentheses() {
        let (name, fields) = split_process_stat(
            "321 (worker (batch)) S 11 0 0 0 0 0 0 0 0 0 7 8 0 0 0 0 0 0 900 0 42",
        )
        .unwrap();

        assert_eq!(name, "worker (batch)");
        assert_eq!(fields[1], "11");
        assert_eq!(fields[11], "7");
        assert_eq!(fields[12], "8");
        assert_eq!(fields[19], "900");
        assert_eq!(fields[21], "42");
    }

    #[test]
    fn cpu_total_uses_host_fields_once() {
        let (total, idle) = cpu_counters("cpu 1 2 3 4 5 6 7 8 900 1000");

        assert_eq!(total, 36);
        assert_eq!(idle, 9);
    }
}

File: apps/server/src/processes/linux.ts
import { readdirSync, readFileSync, readlinkSync, statfsSync } from 'node:fs';
import { availableParallelism, totalmem } from 'node:os';
import { parseEnvironment, type CommandLine, type ProcessSampler, type RawProcess, type RawSample } from './sampler.ts';

/* USER_HZ: the unit of every tick in `/proc`, fixed at 100 for userspace whatever the kernel runs at. */
const CLOCK_TICKS = 100;

export interface ProcStat {
    name: string;
    ppid: number;
    /* utime plus stime, in ticks. */
    cpuTicks: number;
    /* Ticks since boot. */
    startTicks: number;
}

/* `/proc/<pid>/stat`. The name sits in parentheses and may hold spaces and parentheses of its own, so the fields count from the last one. */
export const parseStat = (text: string): ProcStat | null => {
    const open = text.indexOf('(');
    const close = text.lastIndexOf(')');
    if (open === -1 || close === -1) {
        return null;
    }
    // Field 3 (state) is index 0 here.
    const fields = text
        .slice(close + 2)
        .trim()
        .split(' ');
    const ppid = Number(fields[1]);
    const cpuTicks = Number(fields[11]) + Number(fields[12]);
    const startTicks = Number(fields[19]);
    if (!Number.isFinite(ppid) || !Number.isFinite(cpuTicks) || !Number.isFinite(startTicks)) {
        return null;
    }
    return { name: text.slice(open + 1, close), ppid, cpuTicks, startTicks };
};

/* `Key: value` lines as `/proc/<pid>/status`, `/proc/<pid>/io` and `/proc/meminfo` write them, the first number of each value. */
export const parseKeyValues = (text: string): Map<string, number> => {
    const values = new Map<string, number>();
    for (const line of text.split('\n')) {
        const colon = line.indexOf(':');
        if (colon === -1) {
            continue;
        }
        const number = Number.parseInt(line.slice(colon + 1).trim(), 10);
        if (Number.isFinite(number)) {
            values.set(line.slice(0, colon), number);
        }
    }
    return values;
};

/* The first line of `/proc/stat`. Busy is everything but idle and iowait. */
export const parseCpuLine = (text: string): { busy: number; total: number } | null => {
    const line = text.split('\n').find((entry) => entry.startsWith('cpu '));
    if (!line) {
        return null;
    }
    const ticks = line.trim().split(/\s+/).slice(1).map(Number);
    const total = ticks.slice(0, 8).reduce((sum, value) => sum + (Number.isFinite(value) ? value : 0), 0);
    const idle = (ticks[3] ?? 0) + (ticks[4] ?? 0);
    return { busy: total - idle, total };
};

const read = (path: string): string | null => {
    try {
        return readFileSync(path, 'latin1');
    } catch {
        return null;
    }
};

export class LinuxSampler implements ProcessSampler {
    private readonly home: string;
    private readonly bootSeconds: number;

    constructor(home: string) {
        this.home = home;
        const btime = /^btime (\d+)$/m.exec(read('/proc/stat') ?? '');
        this.bootSeconds = btime ? Number(btime[1]) : Math.round(Date.now() / 1000 - this.uptimeSeconds());
    }

    sample(): RawSample {
        const awakeMs = performance.timeOrigin + performance.now();
        const at = Date.now();
        const processes: RawProcess[] = [];
        for (const entry of readdirSync('/proc')) {
            const pid = Number(entry);
            if (Number.isInteger(pid) && pid > 0) {
                const process = this.read(pid);
                if (process !== null) {
                    processes.push(process);
                }
            }
        }
        return {
            at,
            awakeMs,
            // CLOCK_BOOTTIME keeps counting through a suspend and the monotonic clock does not.
            asleepMs: this.uptimeSeconds() * 1000 - awakeMs,
            processes,
            machine: this.machine()
        };
    }

    inspect(pid: number): { startTime: number; uid: number } | null {
        const stat = parseStat(read(`/proc/${pid}/stat`) ?? '');
        const status = parseKeyValues(read(`/proc/${pid}/status`) ?? '');
        const uid = status.get('Uid');
        if (stat === null || uid === undefined) {
            return null;
        }
        return { startTime: this.startTimeOf(stat), uid };
    }

    commandLine(pid: number): CommandLine | null {
        const args = read(`/proc/${pid}/cmdline`);
        const environ = read(`/proc/${pid}/environ`);
        if (args === null || environ === null) {
            return null;
        }
        return { args: args.split('\0').filter((arg) => arg !== ''), env: parseEnvironment(environ.split('\0')) };
    }

    private uptimeSeconds(): number {
        return Number.parseFloat(read('/proc/uptime') ?? '0');
    }

    private startTimeOf(stat: ProcStat): number {
        return this.bootSeconds * 1e6 + (stat.startTicks * 1e6) / CLOCK_TICKS;
    }

    private read(pid: number): RawProcess | null {
        const stat = parseStat(read(`/proc/${pid}/stat`) ?? '');
        if (stat === null) {
            return null;
        }
        const status = parseKeyValues(read(`/proc/${pid}/status`) ?? '');
        let path: string | null = null;
        try {
            path = readlinkSync(`/proc/${pid}/exe`);
        } catch {
            // Another user's executable, or a kernel thread that has none.
        }
        // `io` is only readable for processes of this user; the CPU of the others still is.
        const io = read(`/proc/${pid}/io`);
        const counters = io === null ? null : parseKeyValues(io);
        const rss = status.get('VmRSS');
        return {
            pid,
            ppid: stat.ppid,
            uid: status.get('Uid') ?? -1,
            startTime: this.startTimeOf(stat),
            name: stat.name,
            path,
            readable: true,
            cpuNs: (stat.cpuTicks * 1e9) / CLOCK_TICKS,
            memory: rss === undefined ? null : rss * 1024,
            diskRead: counters?.get('read_bytes') ?? null,
            diskWrite: counters?.get('write_bytes') ?? null
        };
    }

    private machine(): RawSample['machine'] {
        const cpu = parseCpuLine(read('/proc/stat') ?? '');
        const memory = parseKeyValues(read('/proc/meminfo') ?? '');
        const total = memory.get('MemTotal');
        const available = memory.get('MemAvailable');
        let diskFree: number | null = null;
        let diskTotal: number | null = null;
        try {
            const volume = statfsSync(this.home);
            diskFree = volume.bavail * volume.bsize;
            diskTotal = volume.blocks * volume.bsize;
        } catch {
            // The rest of the reading still stands.
        }
        return {
            cores: availableParallelism(),
            cpuBusy: cpu?.busy ?? null,
            cpuTotal: cpu?.total ?? null,
            memoryUsed: total !== undefined && available !== undefined ? (total - available) * 1024 : null,
            memoryTotal: total !== undefined ? total * 1024 : totalmem(),
            diskFree,
            diskTotal
        };
    }
}
